---
title: "Análisis de Enlaces Web"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(rvest)
```



# Sección 1 - Página Principal

```{r}
# URL de la página web que quieres obtener
url <- "http://www.uusmb.unam.mx/ideamex/elisa.azuara141845/#Analisis"

# Realizar la solicitud y obtener el contenido de la página
pagina <- read_html(url)

# Extraer enlaces de la página
enlaces <- pagina %>% html_nodes("a") %>% html_attr("href")

# Mostrar los enlaces obtenidos
print(enlaces)

# Extraer enlaces y direcciones completas de la página
enlaces <- pagina %>% html_nodes("a")
texto_enlaces <- html_text(enlaces)
atributos_enlaces <- enlaces %>% html_attr("href")
direcciones_enlaces <- URLencode(paste(url, atributos_enlaces, sep = ""))

# Crear un data frame con los enlaces y sus detalles
data_enlaces <- data.frame(
  Texto = texto_enlaces,
  Enlace = atributos_enlaces,
  Direccion = direcciones_enlaces
)

# Imprimir el data frame con los enlaces y sus detalles
print(data_enlaces)

# Convertir el contenido HTML en texto
codigo_html <- as.character(pagina)

# Mostrar el código HTML
cat(codigo_html)
```

Data Analysis module allows the user to have a preliminary analysis for quality control based on the data distribution per sample, using plots below

[![CPM, Counts Per Million Plot](DataAnalysis_Results/AllConditions_CPMPlot.png)](DataAnalysis_Results/AllConditions_CPMPlot.pdf)
*Fig. 1. CPM, Counts Per Million Plot*

The Count per million (CPM) plot shows the number of genes within each sample having more than 0 CPM, or more than 1, 2, 5 and 10 CPM. This plot could help the user decide the threshold to remove genes that appear to be very lowly expressed in any of the experimental conditions.

[![Boxplot with pseudo counts](DataAnalysis_Results/AllConditions_BoxPlotLog2.png)](DataAnalysis_Results/AllConditions_BoxPlotLog2.pdf)
*Fig. 2. Boxplot with pseudo counts*

Samples boxplot provides an easy way to visualize the distribution of pseudocounts in each sample, since it shows statistical measures such as median, quartiles, minimum and maximum. Whiskers are also drawn extending beyond each end of the box with points beyond the whiskers typically indicating outliers.

[![Density plot](DataAnalysis_Results/AllConditions_DensitiesPlot.png)](DataAnalysis_Results/AllConditions_DensitiesPlot.pdf)
*Fig. 3. Density plot*

Pseudocounts distributions can also be summarized by means of a density plot. Density plot provides more detail by enabling, for example, the detection of dissimilarity in replicates' behavior.

[![Multi-dimensional scaling plot](DataAnalysis_Results/AllConditions_MDSPlot.png)](DataAnalysis_Results/AllConditions_MDSPlot.pdf)
*Fig. 4. Multi-dimensional scaling plot*

Multidimensional Scaling (MDS) is a technique that is used to create a visual representation of the pattern of proximities (similarities, dissimilarities, or distances) among a set of objects. In the context of RNA-Seq analysis, MDS plot shows variation among RNA-seq samples, distance between sample labels indicates dissimilarity.

[![Principal component plot](DataAnalysis_Results/AllConditions_PCAPlot.png)](DataAnalysis_Results/AllConditions_PCAPlot.pdf)
*Fig. 5. Principal component plot*

This type of plot is useful for visualizing the overall effect of experimental covariates and batch effects. In the context of RNA-Seq analysis, PCA essentially clusters samples by groups of the most significantly deregulated genes. Clustering first by the most significant group, then by progressively less significant groups.

[![Boxplot of standardized counts with TMM](DataAnalysis_Results/AllConditions_BoxPlotNorm.png)](DataAnalysis_Results/AllConditions_BoxPlotNorm.pdf)
*Fig. 6. Boxplot of standardized counts with TMM*

[![Multi-dimensional scaling plot of standardized counts with TMM](DataAnalysis_Results/AllConditions_MDSPlotNorm.png)](DataAnalysis_Results/AllConditions_MDSPlotNorm.pdf)
*Fig. 7. Multi-dimensional scaling plot of standardized counts with TMM*

[![Principal component plot of standardized counts with TMM](DataAnalysis_Results/AllConditions_PCAPlotNorm.png)](DataAnalysis_Results/AllConditions_PCAPlotNorm.pdf)
*Fig. 8. Principal component plot of standardized counts with TMM*

This type of graph is useful for visualizing the overall effect of experimental covariances and their batch effects. When a pair of samples, under the same condition, tend to group together, a lot effect can be suspected.

From this point, the same graphs are replicated, but they show the raw counts normalized with the TMM method (Trimmed Mean of M-values), (1) Robinson.

**References:**

Mark D Robinson, Alicia Oshlack.
"A scaling normalization method for differential expression analysis of RNA-seq data."
Genome Biology 2010, 11:R25. DOI: 10.1186/gb-2010-11-3-r25.


# Sección 2 - EdgeR.php
```{r}

# URL de la página web que quieres obtener
url <- "http://www.uusmb.unam.mx/ideamex/elisa.azuara141845/EdgeR.php"

# Realizar la solicitud y obtener el contenido de la página
pagina <- read_html(url)

# Extraer enlaces de la página
enlaces <- pagina %>% html_nodes("a") %>% html_attr("href")

# Mostrar los enlaces obtenidos
print(enlaces)

# Extraer enlaces y direcciones completas de la página
enlaces <- pagina %>% html_nodes("a")
texto_enlaces <- html_text(enlaces)
atributos_enlaces <- enlaces %>% html_attr("href")
direcciones_enlaces <- URLencode(paste(url, atributos_enlaces, sep = ""))

# Crear un data frame con los enlaces y sus detalles
data_enlaces <- data.frame(
  Texto = texto_enlaces,
  Enlace = atributos_enlaces,
  Direccion = direcciones_enlaces
)

# Imprimir el data frame con los enlaces y sus detalles
print(data_enlaces)

# Convertir el contenido HTML en texto
codigo_html <- as.character(pagina)

# Mostrar el código HTML
cat(codigo_html)
```

In this section, we present the results of the differential expression analysis generated by the edgeR method. (1)

## Result Files

The following text files contain the results of differential expression analysis for various comparisons:

- [<code>pEhExvsU2AF84_pval.txt</code>](edgeR_Results/pEhExvsU2AF84_pval.txt)
- [<code>pEhExvsEhMyb10.txt</code>](edgeR_Results/pEhExvsEhMyb10.txt)
- [<code>pEhExvsUmasM.txt</code>](edgeR_Results/pEhExvsUmasM.txt)
- [<code>pEhExvsCDC5.txt</code>](edgeR_Results/pEhExvsCDC5.txt)
- [<code>pEhExvsEhMyb10_Abundances.txt</code>](edgeR_Results/pEhExvsEhMyb10_Abundances.txt)
- [<code>pEhExvsU2AF84_TOP.txt</code>](edgeR_Results/pEhExvsU2AF84_TOP.txt)
- [<code>pEhExvsUmasM_logFC.txt</code>](edgeR_Results/pEhExvsUmasM_logFC.txt)
- [<code>pEhExvsCmasM.txt</code>](edgeR_Results/pEhExvsCmasM.txt)
- [<code>pEhExvsCmasM_Abundances.txt</code>](edgeR_Results/pEhExvsCmasM_Abundances.txt)
- [<code>pEhExvsUmasM_TOP.txt</code>](edgeR_Results/pEhExvsUmasM_TOP.txt)
- [<code>pEhExvsUmasM_pval.txt</code>](edgeR_Results/pEhExvsUmasM_pval.txt)
- [<code>pEhExvsU2AF84_Abundances.txt</code>](edgeR_Results/pEhExvsU2AF84_Abundances.txt)
- [<code>pEhExvsU2AF84_logFC.txt</code>](edgeR_Results/pEhExvsU2AF84_logFC.txt)
- [<code>pEhExvsCDC5_logFC.txt</code>](edgeR_Results/pEhExvsCDC5_logFC.txt)
- [<code>pEhExvsCmasM_logFC.txt</code>](edgeR_Results/pEhExvsCmasM_logFC.txt)
- [<code>pEhExvsCDC5_Abundances.txt</code>](edgeR_Results/pEhExvsCDC5_Abundances.txt)
- [<code>pEhExvsUmasM_Abundances.txt</code>](edgeR_Results/pEhExvsUmasM_Abundances.txt)
- [<code>pEhExvsEhMyb10_TOP.txt</code>](edgeR_Results/pEhExvsEhMyb10_TOP.txt)
- [<code>pEhExvsCDC5_pval.txt</code>](edgeR_Results/pEhExvsCDC5_pval.txt)
- [<code>pEhExvsEhMyb10_pval.txt</code>](edgeR_Results/pEhExvsEhMyb10_pval.txt)
- [<code>pEhExvsCmasM_TOP.txt</code>](edgeR_Results/pEhExvsCmasM_TOP.txt)
- [<code>pEhExvsEhMyb10_logFC.txt</code>](edgeR_Results/pEhExvsEhMyb10_logFC.txt)
- [<code>pEhExvsCDC5_TOP.txt</code>](edgeR_Results/pEhExvsCDC5_TOP.txt)
- [<code>pEhExvsCmasM_pval.txt</code>](edgeR_Results/pEhExvsCmasM_pval.txt)
- [<code>pEhExvsU2AF84.txt</code>](edgeR_Results/pEhExvsU2AF84.txt)



The text files can be opened in Excel and contain the result of the differential expression analysis for all genes with significant counts. The meaning of each column is as follows:

1. **id**: The gene ID (gene, transcript, etc.)
2. **log2FoldChange**: The log fold-change between conditions being tested
3. **logCPM**: Average log2-counts per million, averaged over all libraries
4. **p value**: The statistical significance of the change
5. **FDR**: False Discovery Rate adjusted for multiple testing with the Benjamini-Hochberg procedure

The last columns correspond to:

6. **raw counts**: Raw counts for all samples
7. **normalized counts**: Normalized counts for all samples
8. **Regulation**: Gene regulation, indicating how and in which condition the gene was expressed

## Subgroup Analysis

Additional analysis files:

- [<code>pEhExvsU2AF84_TOP.txt</code>](edgeR_Results/pEhExvsU2AF84_TOP.txt)
- [<code>pEhExvsUmasM_TOP.txt</code>](edgeR_Results/pEhExvsUmasM_TOP.txt)
- [<code>pEhExvsEhMyb10_TOP.txt</code>](edgeR_Results/pEhExvsEhMyb10_TOP.txt)
- [<code>pEhExvsCmasM_TOP.txt</code>](edgeR_Results/pEhExvsCmasM_TOP.txt)
- [<code>pEhExvsCDC5_TOP.txt</code>](edgeR_Results/pEhExvsCDC5_TOP.txt)

These "TOP.txt" files represent subgroup analyses and contain results only for differentially expressed genes based on the established cutoff parameters.

- [<code>pEhExvsCmasM_Intersect.txt</code>](edgeR_Results/pEhExvsCmasM_Intersect.txt)

The following files have genes that it was reported as differential expressed for more than one method,(EdgeR + other)

## Gene Expression Plots

In the context of RNA-Seq analysis, MDS plot shows variation among RNA-seq samples, with the distance between sample labels indicating dissimilarity. When the experiment is well controlled, the greatest sources of variation should correspond to the treatments/groups of interest.

The Smear plot visualizes the results of a DE analysis, showing the log-fold change against log-counts per million. Differentially expressed features are highlighted in red.

The Volcano plot summarizes both fold-change and p-values. It is a scatter-plot of the negative log10-transformed p-values against the log2 fold change. Features declared as differentially expressed are highlighted in red.


[![Imagen 1](edgeR_Results/pEhExvsU2AF84_plotVolcano.png)](edgeR_Results/pEhExvsU2AF84_plotVolcano.pdf){:target="_blank"}
[![Imagen 2](edgeR_Results/pEhExvsUmasM_plotSmear.png)](edgeR_Results/pEhExvsUmasM_plotSmear.pdf){:target="_blank"}
[![Imagen 3](edgeR_Results/pEhExvsCDC5_plotVolcano.png)](edgeR_Results/pEhExvsCDC5_plotVolcano.pdf){:target="_blank"}
[![Imagen 4](edgeR_Results/pEhExvsEhMyb10_plotMDS.png)](edgeR_Results/pEhExvsEhMyb10_plotMDS.pdf){:target="_blank"}
[![Imagen 5](edgeR_Results/pEhExvsCDC5_plotSmear.png)](edgeR_Results/pEhExvsCDC5_plotSmear.pdf){:target="_blank"}
[![Imagen 6](edgeR_Results/pEhExvsU2AF84_plotSmear.png)](edgeR_Results/pEhExvsU2AF84_plotSmear.pdf){:target="_blank"}
[![Imagen 7](edgeR_Results/pEhExvsCmasM_plotSmear.png)](edgeR_Results/pEhExvsCmasM_plotSmear.pdf){:target="_blank"}
[![Imagen 8](edgeR_Results/pEhExvsUmasM_plotMDS.png)](edgeR_Results/pEhExvsUmasM_plotMDS.pdf){:target="_blank"}
[![Imagen 9](edgeR_Results/pEhExvsEhMyb10_plotVolcano.png)](edgeR_Results/pEhExvsEhMyb10_plotVolcano.pdf){:target="_blank"}
[![Imagen 10](edgeR_Results/pEhExvsCmasM_plotVolcano.png)](edgeR_Results/pEhExvsCmasM_plotVolcano.pdf){:target="_blank"}
[![Imagen 11](edgeR_Results/pEhExvsCmasM_plotMDS.png)](edgeR_Results/pEhExvsCmasM_plotMDS.pdf){:target="_blank"}
[![Imagen 12](edgeR_Results/pEhExvsU2AF84_plotMDS.png)](edgeR_Results/pEhExvsU2AF84_plotMDS.pdf){:target="_blank"}
[![Imagen 13](edgeR_Results/pEhExvsEhMyb10_plotSmear.png)](edgeR_Results/pEhExvsEhMyb10_plotSmear.pdf){:target="_blank"}
[![Imagen 14](edgeR_Results/pEhExvsUmasM_plotVolcano.png)](edgeR_Results/pEhExvsUmasM_plotVolcano.pdf){:target="_blank"}

In the context of RNA-Seq analysis, MDS plot shows variation among RNA-seq samples, distance between sample labels indicates dissimilarity. If the experiment is well controlled and has worked well, what we hope to see is that the greatest sources of variation in the data are the treatments/groups we are interested in

The Smear plot allows to visualise the results of a DE analysis. In a similar manner to the MA-plot, this plot shows the log-fold change against log-counts per million, with features declared as differentially expressed highlighted in red.

The Volcanoplot is a graph that summarizes both fold-change and the p-value. It is a scatter-plot of the negative log10-transformed p-values from the gene-specific test (on the y-axis) against the log2 fold change (on the x-axis). This results in datapoints with low p-values (highly significant) appearing towards the top of the plot. The log2 of the fold-change is used so that changes in both directions (up and down) appear equidistant from the center (Gonzalez 2014). Features declared as differentially expressed are highlighted in red.

# Sección 3 - DESeq2.php
```{r}

# URL de la página web que quieres obtener
url <- "http://www.uusmb.unam.mx/ideamex/elisa.azuara141845/DESeq2.php"

# Realizar la solicitud y obtener el contenido de la página
pagina <- read_html(url)

# Extraer enlaces de la página
enlaces <- pagina %>% html_nodes("a") %>% html_attr("href")

# Mostrar los enlaces obtenidos
print(enlaces)

# Extraer enlaces y direcciones completas de la página
enlaces <- pagina %>% html_nodes("a")
texto_enlaces <- html_text(enlaces)
atributos_enlaces <- enlaces %>% html_attr("href")
direcciones_enlaces <- URLencode(paste(url, atributos_enlaces, sep = ""))

# Crear un data frame con los enlaces y sus detalles
data_enlaces <- data.frame(
  Texto = texto_enlaces,
  Enlace = atributos_enlaces,
  Direccion = direcciones_enlaces
)

# Imprimir el data frame con los enlaces y sus detalles
print(data_enlaces)

# Convertir el contenido HTML en texto
codigo_html <- as.character(pagina)

# Mostrar el código HTML
cat(codigo_html)
```

# Sección 4 - limma.php
```{r}

# URL de la página web que quieres obtener
url <- "http://www.uusmb.unam.mx/ideamex/elisa.azuara141845/limma.php"

# Realizar la solicitud y obtener el contenido de la página
pagina <- read_html(url)

# Extraer enlaces de la página
enlaces <- pagina %>% html_nodes("a") %>% html_attr("href")

# Mostrar los enlaces obtenidos
print(enlaces)

# Extraer enlaces y direcciones completas de la página
enlaces <- pagina %>% html_nodes("a")
texto_enlaces <- html_text(enlaces)
atributos_enlaces <- enlaces %>% html_attr("href")
direcciones_enlaces <- URLencode(paste(url, atributos_enlaces, sep = ""))

# Crear un data frame con los enlaces y sus detalles
data_enlaces <- data.frame(
  Texto = texto_enlaces,
  Enlace = atributos_enlaces,
  Direccion = direcciones_enlaces
)

# Imprimir el data frame con los enlaces y sus detalles
print(data_enlaces)

# Convertir el contenido HTML en texto
codigo_html <- as.character(pagina)

# Mostrar el código HTML
cat(codigo_html)
```

# Sección 5 - NOISeq.php
```{r}
# URL de la página web que quieres obtener
url <- "http://www.uusmb.unam.mx/ideamex/elisa.azuara141845/NOISeq.php"

# Realizar la solicitud y obtener el contenido de la página
pagina <- read_html(url)

# Extraer enlaces de la página
enlaces <- pagina %>% html_nodes("a") %>% html_attr("href")

# Mostrar los enlaces obtenidos
print(enlaces)

# Extraer enlaces y direcciones completas de la página
enlaces <- pagina %>% html_nodes("a")
texto_enlaces <- html_text(enlaces)
atributos_enlaces <- enlaces %>% html_attr("href")
direcciones_enlaces <- URLencode(paste(url, atributos_enlaces, sep = ""))

# Crear un data frame con los enlaces y sus detalles
data_enlaces <- data.frame(
  Texto = texto_enlaces,
  Enlace = atributos_enlaces,
  Direccion = direcciones_enlaces
)

# Imprimir el data frame con los enlaces y sus detalles
print(data_enlaces)

# Convertir el contenido HTML en texto
codigo_html <- as.character(pagina)

# Mostrar el código HTML
cat(codigo_html)
```

# Sección 6 - VennDiagram.php
```{r}

# URL de la página web que quieres obtener
url <- "http://www.uusmb.unam.mx/ideamex/elisa.azuara141845/VennDiagram.php"

# Realizar la solicitud y obtener el contenido de la página
pagina <- read_html(url)

# Extraer enlaces de la página
enlaces <- pagina %>% html_nodes("a") %>% html_attr("href")

# Mostrar los enlaces obtenidos
print(enlaces)

# Extraer enlaces y direcciones completas de la página
enlaces <- pagina %>% html_nodes("a")
texto_enlaces <- html_text(enlaces)
atributos_enlaces <- enlaces %>% html_attr("href")
direcciones_enlaces <- URLencode(paste(url, atributos_enlaces, sep = ""))

# Crear un data frame con los enlaces y sus detalles
data_enlaces <- data.frame(
  Texto = texto_enlaces,
  Enlace = atributos_enlaces,
  Direccion = direcciones_enlaces
)

# Imprimir el data frame con los enlaces y sus detalles
print(data_enlaces)

# Convertir el contenido HTML en texto
codigo_html <- as.character(pagina)

# Mostrar el código HTML
cat(codigo_html)
```

